io-sim/src/Control/Monad/IOSimPOR/Internal.hs

```diff
@@ -0,0 +1,1573 @@
+
+unblockAllThreadsFromTVar :: TVar s a -> ST s ()
+unblockAllThreadsFromTVar TVar{tvarBlocked} = do
+    writeSTRef tvarBlocked ([], Set.empty)
+
+-- | For each TVar written to in a transaction (in order) collect the threads
+-- that blocked on each one (in order).
+--
+-- Also, for logging purposes, return an association between the threads and
+-- the var writes that woke them.
+--
+threadsUnblockedByWrites :: [SomeTVar s]
+                         -> ST s ([ThreadId], Map ThreadId (Set (Labelled TVarId)))
+threadsUnblockedByWrites written = do
+  tidss <- sequence
+             [ (,) <$> labelledTVarId tvar <*> readTVarBlockedThreads tvar
+             | SomeTVar tvar <- written ]
+  -- Threads to wake up, in wake up order, annotated with the vars written that
+  -- caused the unblocking.
+  -- We reverse the individual lists because the tvarBlocked is used as a stack
+  -- so it is in order of last written, LIFO, and we want FIFO behaviour.
+  let wakeup = ordNub [ tid | (_vid, tids) <- tidss, tid <- reverse tids ]
+      wokeby = Map.fromListWith Set.union
+                                [ (tid, Set.singleton vid)
+                                | (vid, tids) <- tidss
+                                , tid <- tids ]
+  return (wakeup, wokeby)
+
+ordNub :: Ord a => [a] -> [a]
+ordNub = go Set.empty
+  where
+    go !_ [] = []
+    go !s (x:xs)
+      | x `Set.member` s = go s xs
+      | otherwise        = x : go (Set.insert x s) xs
+
+-- Effects
+
+data Effect = Effect {
+    effectReads  :: !(Set TVarId),
+    effectWrites :: !(Set TVarId),
+    effectForks  :: !(Set ThreadId),
+    effectLiftST :: !Bool,
+    effectThrows :: ![ThreadId],
+    effectWakeup :: ![ThreadId]
+  }
+  deriving (Eq, Show)
+
+instance Semigroup Effect where
+  Effect r w s b ts wu <> Effect r' w' s' b' ts' wu' =
+    Effect (r<>r') (w<>w') (s<>s') (b||b') (ts++ts') (wu++wu')
+
+instance Monoid Effect where
+  mempty = Effect Set.empty Set.empty Set.empty False [] []
+
+-- readEffect :: SomeTVar s -> Effect
+-- readEffect r = mempty{effectReads = Set.singleton $ someTvarId r }
+
+readEffects :: [SomeTVar s] -> Effect
+readEffects rs = mempty{effectReads = Set.fromList (map someTvarId rs)}
+
+-- writeEffect :: SomeTVar s -> Effect
+-- writeEffect r = mempty{effectWrites = Set.singleton $ someTvarId r }
+
+writeEffects :: [SomeTVar s] -> Effect
+writeEffects rs = mempty{effectWrites = Set.fromList (map someTvarId rs)}
+
+forkEffect :: ThreadId -> Effect
+forkEffect tid = mempty{effectForks = Set.singleton $ tid}
+
+liftSTEffect :: Effect
+liftSTEffect = mempty{ effectLiftST = True }
+
+throwToEffect :: ThreadId -> Effect
+throwToEffect tid = mempty{ effectThrows = [tid] }
+
+wakeupEffects :: [ThreadId] -> Effect
+wakeupEffects tids = mempty{effectWakeup = tids}
+
+someTvarId :: SomeTVar s -> TVarId
+someTvarId (SomeTVar r) = tvarId r
+
+onlyReadEffect :: Effect -> Bool
+onlyReadEffect e@Effect { effectReads } = e == mempty { effectReads }
```

Comment:
Or: onlyReadEffect e = e { effectReads = effectReads mempty } == mempty
It would be nice to avoid repeating Set.empty, in case in the future somebody replaces effectReads by a different type.